packages/expo-module-autolinking/src/autolinking.ts

```diff
@@ -0,0 +1,200 @@
+        continue;
+      }
+
+      const currentRevision: PackageRevision = {
+        path: packagePath,
+        version,
+      };
+
+      if (!results[name]) {
+        // The revision that was found first will be the main one.
+        // An array of duplicates is needed only here.
+        results[name] = { ...currentRevision, duplicates: [] };
+      } else if (
+        results[name].path !== packagePath &&
+        results[name].duplicates?.every(({ path }) => path !== packagePath)
+      ) {
+        results[name].duplicates?.push(currentRevision);
+      }
+    }
+  }
+  return results;
+}
+
+/**
+ * Merges autolinking options from different sources (the later the higher priority)
+ * - options defined in package.json's `expoModules` field
+ * - platform-specific options from the above (e.g. `expoModules.ios`)
+ * - options provided to the CLI command
+ */
+export async function mergeLinkingOptionsAsync<OptionsType extends SearchOptions>(
+  providedOptions: OptionsType
+): Promise<OptionsType> {
+  const packageJsonPath = await findPackageJsonPathAsync();
+  const packageJson = packageJsonPath ? require(packageJsonPath) : {};
+  const baseOptions = packageJson.expo?.autolinking;
+  const platformOptions = providedOptions.platform && baseOptions?.[providedOptions.platform];
+  const finalOptions = Object.assign(
+    {},
+    baseOptions,
+    platformOptions,
+    providedOptions
+  ) as OptionsType;
+
+  // Makes provided paths absolute or falls back to default paths if none was provided.
+  finalOptions.searchPaths = await resolveSearchPathsAsync(finalOptions.searchPaths, process.cwd());
+
+  return finalOptions;
+}
+
+/**
+ * Verifies the search results by checking whether there are no duplicates.
+ */
+export function verifySearchResults(searchResults: SearchResults): number {
+  const cwd = process.cwd();
+  const relativePath: (pkg: PackageRevision) => string = pkg => path.relative(cwd, pkg.path);
+  let counter = 0;
+
+  for (const moduleName in searchResults) {
+    const revision = searchResults[moduleName];
+
+    if (revision.duplicates?.length) {
+      console.warn(`⚠️  Found multiple revisions of ${chalk.green(moduleName)}`);
+      console.log(` - ${chalk.magenta(relativePath(revision))} (${chalk.cyan(revision.version)})`);
+
+      for (const duplicate of revision.duplicates) {
+        console.log(` - ${chalk.gray(relativePath(duplicate))} (${chalk.gray(duplicate.version)})`);
+      }
+      counter++;
+    }
+  }
+  if (counter > 0) {
+    console.warn(
+      '⚠️  Please get rid of multiple revisions as it may introduce some side effects or compatibility issues'
+    );
+  }
+  return counter;
+}
+
+/**
+ * Resolves search results to a list of platform-specific configuration.
+ */
+export async function resolveModulesAsync(
+  searchResults: SearchResults,
+  options: ResolveOptions
+): Promise<ModuleDescriptor[]> {
+  const platformLinking = require(`./platforms/${options.platform}`);
```

Comment:
I'm giving you some extra points for leveraging strategy pattern this way here 😂